assert no duplicate yaml keys in codegen

bdhirsh opened the pull request:
The codegen should error if it sees two yaml entries with the same key. The default behavior of python's yaml loader is to overwrite duplicate keys with the new value (see https://github.com/yaml/pyyaml/issues/165).

This would have caught a nasty bug that showed up in https://github.com/pytorch/pytorch/pull/66225/files#r723796194.

I tested it on that linked PR, to confirm that it errors correctly (and gives the line number containing the duplicate).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66238

Differential Revision: [D31464585](https://our.internmc.facebook.com/intern/diff/D31464585)

bdhirsh committed:
assert no duplicate yaml keys in codegen

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31464585).

ezyang reviewed (approved):
This probably regresses YAML loading performance but meh WORTH IT

bdhirsh committed:
Update on "assert no duplicate yaml keys in codegen"

The codegen should error if it sees two yaml entries with the same key. The default behavior of python's yaml loader is to overwrite duplicate keys with the new value (see https://github.com/yaml/pyyaml/issues/165).

This would have caught a nasty bug that showed up in https://github.com/pytorch/pytorch/pull/66225/files#r723796194.

I tested it on that linked PR, to confirm that it errors correctly (and gives the line number containing the duplicate).


Differential Revision: [D31464585](https://our.internmc.facebook.com/intern/diff/D31464585)

[ghstack-poisoned]

bdhirsh committed:
Update on "assert no duplicate yaml keys in codegen"

The codegen should error if it sees two yaml entries with the same key. The default behavior of python's yaml loader is to overwrite duplicate keys with the new value (see https://github.com/yaml/pyyaml/issues/165).

This would have caught a nasty bug that showed up in https://github.com/pytorch/pytorch/pull/66225/files#r723796194.

I tested it on that linked PR, to confirm that it errors correctly (and gives the line number containing the duplicate).


Differential Revision: [D31464585](https://our.internmc.facebook.com/intern/diff/D31464585)

[ghstack-poisoned]

albanD reviewed (approved):
lint is broken?

bdhirsh committed:
Update on "assert no duplicate yaml keys in codegen"

The codegen should error if it sees two yaml entries with the same key. The default behavior of python's yaml loader is to overwrite duplicate keys with the new value (see https://github.com/yaml/pyyaml/issues/165).

This would have caught a nasty bug that showed up in https://github.com/pytorch/pytorch/pull/66225/files#r723796194.

I tested it on that linked PR, to confirm that it errors correctly (and gives the line number containing the duplicate).


Differential Revision: [D31464585](https://our.internmc.facebook.com/intern/diff/D31464585)

[ghstack-poisoned]

bdhirsh committed:
Update on "assert no duplicate yaml keys in codegen"

The codegen should error if it sees two yaml entries with the same key. The default behavior of python's yaml loader is to overwrite duplicate keys with the new value (see https://github.com/yaml/pyyaml/issues/165).

This would have caught a nasty bug that showed up in https://github.com/pytorch/pytorch/pull/66225/files#r723796194.

I tested it on that linked PR, to confirm that it errors correctly (and gives the line number containing the duplicate).


Differential Revision: [D31464585](https://our.internmc.facebook.com/intern/diff/D31464585)

[ghstack-poisoned]
